Source/Core/VideoCommon/Fifo.cpp

```diff
@@ -413,54 +413,61 @@ bool AtBreakpoint()
 
 void RunGpu()
 {
-	SCPFifoStruct &fifo = CommandProcessor::fifo;
 	const SConfig& param = SConfig::GetInstance();
 
-	// execute GPU
-	if (!param.bCPUThread || s_use_deterministic_gpu_thread)
+	// wake up GPU thread
+	if (param.bCPUThread && !s_use_deterministic_gpu_thread)
+	{
+		s_gpu_mainloop.Wakeup();
+	}
+}
+
+static int RunGpuOnCpu(int ticks)
+{
+	SCPFifoStruct &fifo = CommandProcessor::fifo;
+	bool reset_simd_state = false;
+	int available_ticks = int(ticks * SConfig::GetInstance().fSyncGpuOverclock) + s_sync_ticks.load();
+	while (fifo.bFF_GPReadEnable && fifo.CPReadWriteDistance && !AtBreakpoint() && available_ticks > 0)
 	{
-		bool reset_simd_state = false;
-		while (fifo.bFF_GPReadEnable && fifo.CPReadWriteDistance && !AtBreakpoint() )
+		if (s_use_deterministic_gpu_thread)
 		{
-			if (s_use_deterministic_gpu_thread)
-			{
-				ReadDataFromFifoOnCPU(fifo.CPReadPointer);
-				s_gpu_mainloop.Wakeup();
-			}
-			else
+			ReadDataFromFifoOnCPU(fifo.CPReadPointer);
+			s_gpu_mainloop.Wakeup();
+		}
+		else
+		{
+			if (!reset_simd_state)
 			{
-				if (!reset_simd_state)
-				{
-					FPURoundMode::SaveSIMDState();
-					FPURoundMode::LoadDefaultSIMDState();
-					reset_simd_state = true;
-				}
-				ReadDataFromFifo(fifo.CPReadPointer);
-				s_video_buffer_read_ptr = OpcodeDecoder::Run(DataReader(s_video_buffer_read_ptr, s_video_buffer_write_ptr), nullptr, false);
+				FPURoundMode::SaveSIMDState();
+				FPURoundMode::LoadDefaultSIMDState();
+				reset_simd_state = true;
 			}
+			ReadDataFromFifo(fifo.CPReadPointer);
+			u32 cycles = 0;
+			s_video_buffer_read_ptr = OpcodeDecoder::Run(DataReader(s_video_buffer_read_ptr, s_video_buffer_write_ptr), &cycles, false);
+			available_ticks -= cycles;
+		}
 
-			//DEBUG_LOG(COMMANDPROCESSOR, "Fifo wraps to base");
-
-			if (fifo.CPReadPointer == fifo.CPEnd)
-				fifo.CPReadPointer = fifo.CPBase;
-			else
-				fifo.CPReadPointer += 32;
+		//DEBUG_LOG(COMMANDPROCESSOR, "Fifo wraps to base");
 
-			fifo.CPReadWriteDistance -= 32;
-		}
-		CommandProcessor::SetCPStatusFromGPU();
+		if (fifo.CPReadPointer == fifo.CPEnd)
+			fifo.CPReadPointer = fifo.CPBase;
+		else
+			fifo.CPReadPointer += 32;
 
-		if (reset_simd_state)
-		{
-			FPURoundMode::LoadSIMDState();
-		}
+		fifo.CPReadWriteDistance -= 32;
 	}
+	CommandProcessor::SetCPStatusFromGPU();
 
-	// wake up GPU thread
-	if (param.bCPUThread)
+	if (reset_simd_state)
 	{
-		s_gpu_mainloop.Wakeup();
+		FPURoundMode::LoadSIMDState();
 	}
+
+	available_ticks = std::min(available_ticks, 0);
```

Comment:
Yeah, took me a while to work out exactly what you were doing.

If we have extra cycles (available_ticks greater than zero) we discard them. 

If we used too many cycles, say because we executed a large efb copy which takes 76,800 cycles (available_ticks is less than zero) we need to delay the next execution of this callback until after that operation finishes (plus 10,000 cycles)